File: backup/Dynamic_Programming/python/10_정수삼각형.py
# fast campus 강의
# https://www.acmicpc.net/problem/1932
# 2

N = int(input())
dp = [[0] * (N + 1) for _ in range(N + 1)]

matrix = [[0] * (N + 1) for _ in range(N + 1)]
for i in range(1, N + 1):
    tmp = list(map(int, input().split()))
    for j in range(1, i + 1):
        matrix[i][j] = tmp[j - 1]

for i in range(1, N + 1):
    for j in range(1, i + 1):
        dp[i][j] = max(dp[i - 1][j - 1], dp[i - 1][j]) + matrix[i][j]

print(max(dp[-1]))

File: backup/Dynamic_Programming/c++/10_정수삼각형.cpp
// fast capus 강의
// https://www.acmicpc.net/problem/1932
// 2

#include <iostream>
#include <algorithm>
using namespace std;

int N;
int dp[501][501];

int main() {
	cin >> N;

	for (int i = 1; i <= N; i++) {
		for (int j = 1; j <= i; j++) {
			int x;
			cin >> x;
			dp[i][j] = max(dp[i - 1][j - 1], dp[i - 1][j]) + x;
		}
	}

	cout << *max_element(dp[N] + 1, dp[N] + N + 1);
}